presto-main/src/main/java/com/facebook/presto/operator/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,793 @@
+        this.systemMemoryContext.setBytes(partitionsInitialRetainedSize);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    public PartitionedOutputInfo getInfo()
+    {
+        return partitionFunction.getInfo();
+    }
+
+    @Override
+    public void finish()
+    {
+        finished = true;
+        partitionFunction.flush();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finished && isBlocked().isDone();
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        ListenableFuture<?> blocked = partitionFunction.isFull();
+        return blocked.isDone() ? NOT_BLOCKED : blocked;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished && isBlocked().isDone();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+
+        if (page.getPositionCount() == 0) {
+            return;
+        }
+
+        page = pagePreprocessor.apply(page);
+        partitionFunction.partitionPage(page);
+
+        // TODO: PartitionedOutputOperator reports incorrect output data size #11770
+        operatorContext.recordOutput(page.getSizeInBytes(), page.getPositionCount());
+
+        // We use getSizeInBytes() here instead of getRetainedSizeInBytes() for an approximation of
+        // the amount of memory used by the pageBuilders, because calculating the retained
+        // size can be expensive especially for complex types.
+        long partitionsSizeInBytes = partitionFunction.getSizeInBytes();
+
+        // We also add partitionsInitialRetainedSize as an approximation of the object overhead of the partitions.
+        systemMemoryContext.setBytes(partitionsSizeInBytes + partitionsInitialRetainedSize);
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        return null;
+    }
+
+    private static class PartitionData
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(PartitionData.class).instanceSize();
+        private static final int INITIAL_POSITION_COUNT = 64 * 1024;
+        private static final int DEFAULT_ELEMENT_SIZE_IN_BYTES = 8;
+
+        private final int partition;
+        private final AtomicLong rowsAdded;
+        private final AtomicLong pagesAdded;
+        private final PagesSerde serde;
+        private final Lifespan lifespan;
+        private final int capacity;
+        private final int channelCount;
+
+        private int[] positions = new int[INITIAL_POSITION_COUNT];   // the default positions array for top level BlockEncodingBuffers
+        private int[] rowSizes = new int[INITIAL_POSITION_COUNT];
+        private int positionCount;  // number of positions to be copied for this partition
+        private BlockEncodingBuffers[] blockEncodingBuffers;
+
+        private int bufferedRowCount;
```

Comment:
bufferedRowCount -> bufferedPositionCount

Reply:
Original operator uses rowsAdded. Shall we rename that too?

Reply:
I think the original operator can stay the same. Hopefully, it will be gone soon.